[UnifiedPDF] UnifiedPDFPlugin should start producing cursor updates

aprotyas opened the pull request:
#### 907be252d130e3290d0876282300061f0051f2e7
<pre>
[UnifiedPDF] UnifiedPDFPlugin should start producing cursor updates
<a href="https://bugs.webkit.org/show_bug.cgi?id=265512">https://bugs.webkit.org/show_bug.cgi?id=265512</a>
<a href="https://rdar.apple.com/118925385">rdar://118925385</a>

Reviewed by Simon Fraser.

This commit lays down the plumbing required to start generating cursor
updates from UnifiedPDFPlugin, and is written in anticipation of correct
hit testing to identify areas of interest under the mouse -- see
<a href="https://rdar.apple.com/118550951.">rdar://118550951.</a> Note that this commit is not involved in actually
notifying the web page about cursor updates, since that was already
hooked up in 270863@main.

We introduce the `PDFElementTypes` option set to indicate the presence
of interesting PDF elements under a given point. This mirrors the
`PDFAreaOfInterest` enumeration from PDFKit.

Furthermore, we update `UnifiedPDFPlugin::handleMouseEvent()` to reason
about mouseMove events by passing the event in question to
`UnifiedPDFPlugin::pdfElementTypesForPluginPoint()`. This function is not
implemented yet, but will be done soon, as mentioned above.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::toWebCoreCursorType):
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):

Canonical link: <a href="https://commits.webkit.org/271319@main">https://commits.webkit.org/271319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b64e3cd61e145def50229c705b6fdfe553b91230

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27995 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29299 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30525 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25531 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8631 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4023 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25294 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5385 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24052 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4642 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4818 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31214 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25596 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25495 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31077 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4824 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2988 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28888 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6361 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6716 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -30,15 +30,30 @@
 public:
     static Ref<UnifiedPDFPlugin> create(WebCore::HTMLPlugInElement&);
 
+    enum class ElementUnderMouse : uint16_t {
```
probably this should not be about "mouse" but under a "point"

hortont424 replied:
(or just more generic "type" or something)

smfr replied:
`Element` feels wrong in a web context, since it has a very specific meaning in DOM. Also the "under mouse" part of the name seems wrong; this enum is just a list of types. Maybe `InteractableTypes` or something

aprotyas replied:
Going with `PDFElementTypes` for now

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -30,15 +30,30 @@
     static Ref<UnifiedPDFPlugin> create(WebCore::HTMLPlugInElement&);
 
+    enum class ElementUnderMouse : uint16_t {
+        None = 1 << 0,
```
Why is None bit 1 instead of the lack of bits?

smfr replied:
If this is used in an OptionSet, there should be no zero value.

Cool kids are aligning the `=` these days (making it easier to see mistakes in the values).

aprotyas replied:
None bit 1 was an oversight, fixed. Also aligned the `=` to become a cool kid.

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -363,8 +365,49 @@
+{
+    auto cursor = WebCore::Cursor::Type::Pointer;
+
+    if (elementsUnderMouse.containsAny({ UnifiedPDFPlugin::ElementUnderMouse::Link, UnifiedPDFPlugin::ElementUnderMouse::Control, UnifiedPDFPlugin::ElementUnderMouse::Icon }))
```
this could probably be just a bunch of early returns instead of an if-else chain, right?

aprotyas replied:
Sure, yeah, I think I can re-write like this

```cpp
static WebCore::Cursor::Type toWebCoreCursorType(UnifiedPDFPlugin::ElementsUnderMouse elementsUnderMouse, AltKeyIsActive altKeyIsActive = AltKeyIsActive::No)
{
    if (elementsUnderMouse.containsAny({ UnifiedPDFPlugin::ElementUnderMouse::Link, UnifiedPDFPlugin::ElementUnderMouse::Control, UnifiedPDFPlugin::ElementUnderMouse::Icon }))
        return WebCore::Cursor::Type::Hand;
    if (altKeyIsActive == AltKeyIsActive::Yes)
        return WebCore::Cursor::Type::Hand;
    if (elementsUnderMouse.containsAny({ UnifiedPDFPlugin::ElementUnderMouse::Text, UnifiedPDFPlugin::ElementUnderMouse::TextField }))
        return WebCore::Cursor::Type::IBeam;

    return WebCore::Cursor::Type::Pointer;
}
```

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -363,8 +365,49 @@
+        case WebMouseEventButton::None:
+        {
+            auto altKeyIsActive = event.altKey() ? AltKeyIsActive::Yes : AltKeyIsActive::No;
+            auto elementsUnderMouse = getElementsUnderMouse(event);
```
Point, again

hortont424 replied:
Or call it hit test like we do elsewhere. Lots of options, but "mouse" is definitely too specific

aprotyas replied:
Maybe just `auto hitTestResult = hitTestResultForMouseEvent(event);`?

aprotyas replied:
Changed to `auto pdfElementTypes = pdfElementTypesForPluginPoint(mousePositionInPluginCoordinates);`. The callee was refactored to receive plugin coordinates, rather than events.

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -363,8 +365,49 @@
+        default:
+            return false;
+        }
+    default:
```
this default can just be return false and then the bottom-most return false can be omitted

aprotyas replied:
Fixed

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -119,6 +134,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
 
     RefPtr<WebCore::GraphicsLayer> createGraphicsLayer(const String& name, GraphicsLayer::Type);
 
+    ElementsUnderMouse getElementsUnderMouse(const WebMouseEvent&) const;
```
get implies an out argument, no get prefix

aprotyas replied:
Fixed

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -119,6 +134,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
 
     RefPtr<WebCore::GraphicsLayer> createGraphicsLayer(const String& name, GraphicsLayer::Type);
 
+    ElementsUnderMouse getElementsUnderMouse(const WebMouseEvent&) const;
```
Write this as `interactablesForPluginPoint()` (choose your coordinate system) and provide the "for event" as a wrapper.

aprotyas replied:
Settled at `pdfElementTypesForPluginPoint()`. I've extracted the event position in plugin coordinates for now, but I'll make a wrapper to provide the "for event" bit when we're handling other events too.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -363,8 +365,49 @@
+        mouseMovedInContentArea();
+        switch (event.button()) {
+        case WebMouseEventButton::None:
+        {
```
Goes on the previous line.

aprotyas replied:
Fixed

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -363,8 +365,49 @@
+    return cursor;
+}
+
+UnifiedPDFPlugin::ElementsUnderMouse UnifiedPDFPlugin::getElementsUnderMouse(const WebMouseEvent&) const
```
To avoid name spacing the return value, you can write as:
`auto UnifiedPDFPlugin::getElementsUnderMouse(const WebMouseEvent&) const -> ElementsUnderMouse`

aprotyas replied:
TIL, thanks for the tip

aprotyas committed:
[UnifiedPDF] UnifiedPDFPlugin should start producing cursor updates
https://bugs.webkit.org/show_bug.cgi?id=265512
rdar://118925385

Reviewed by Simon Fraser.

This commit lays down the plumbing required to start generating cursor
updates from UnifiedPDFPlugin, and is written in anticipation of correct
hit testing to identify areas of interest under the mouse -- see
rdar://118550951. Note that this commit is not involved in actually
notifying the web page about cursor updates, since that was already
hooked up in 270863@main.

We introduce the `PDFElementTypes` option set to indicate the presence
of interesting PDF elements under a given point. This mirrors the
`PDFAreaOfInterest` enumeration from PDFKit.

Furthermore, we update `UnifiedPDFPlugin::handleMouseEvent()` to reason
about mouseMove events by passing the event in question to
`UnifiedPDFPlugin::pdfElementTypesForPluginPoint()`. This function is not
implemented yet, but will be done soon, as mentioned above.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::toWebCoreCursorType):
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):

Canonical link: https://commits.webkit.org/271319@main
